File: frontend/billing-frontend/src/app/shared/services/payment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Payment, PaymentRequest } from '../models/models';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class PaymentService {
  private apiUrl = `${environment.apiUrl}/api`;

  constructor(private http: HttpClient) {}

  createPayment(request: PaymentRequest): Observable<Payment> {
    return this.http.post<Payment>(`${this.apiUrl}/payments/create`, request);
  }

  initiatePaytmPayment(invoiceId: number): Observable<Payment> {
    return this.http.post<Payment>(`${this.apiUrl}/gateway/paytm/initiate`, null, {
      params: { invoiceId: invoiceId.toString() }
    });
  }

  getPaymentByInvoiceId(invoiceId: number): Observable<Payment> {
    return this.http.get<Payment>(`${this.apiUrl}/payments/invoice/${invoiceId}`);
  }
}

File: backend/src/main/java/com/example/billing/service/PaymentService.java
package com.example.billing.service;

import com.example.billing.dto.request.PaymentRequest;
import com.example.billing.dto.response.PaymentResponse;
import java.util.Map;

public interface PaymentService {

    PaymentResponse createPayment(PaymentRequest request);

    PaymentResponse initiatePaytmPayment(Long invoiceId);

    PaymentResponse handlePaytmCallback(Map<String, String> callbackParams);

    PaymentResponse getPaymentByInvoiceId(Long invoiceId);
}
